pkg/crds/crds.go

```diff
@@ -0,0 +1,224 @@
+			if k8sErrors.IsNotFound(err) {
+				logging.Info("the server could not find the requested resource: %s", resource)
+				continue
+			}
+			return err
+		}
+
+		if len(items) == 0 {
+			logging.Info("resource %s was not found on the cluster", resource)
+			continue
+		}
+
+		for _, item := range items {
+			logging.Info(fmt.Sprintf("removing resource %s: %s", resource, item.GetName()))
+			err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+			if err != nil {
+				if k8sErrors.IsNotFound(err) {
+					logging.Info(err.Error())
+					continue
+				}
+				return err
+			}
+
+			itemResource, err := clientsets.GetResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, item.GetName(), clusterNamespace)
+			if err != nil {
+				if !k8sErrors.IsNotFound(err) {
+					return err
+				}
+			}
+
+			if itemResource != nil {
+				logging.Info(fmt.Sprintf("resource %q is not yet deleted, applying patch to remove finalizer...", itemResource.GetName()))
+				err = updatingFinalizers(ctx, clientsets, itemResource, resource, clusterNamespace)
+				if err != nil {
+					if k8sErrors.IsNotFound(err) {
+						logging.Info(err.Error())
+						continue
+					}
+					return err
+				}
+
+				err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+				if err != nil {
+					if !k8sErrors.IsNotFound(err) {
+						return err
+					}
+				}
+			}
+
+			itemResource, err = clientsets.GetResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, item.GetName(), clusterNamespace)
+			if err != nil {
+				if !k8sErrors.IsNotFound(err) {
+					return err
+				}
+			}
+
+			logging.Info("resource %s was deleted", item.GetName())
+		}
+	}
+	return nil
+}
+
+func updatingFinalizers(ctx context.Context, clientsets k8sutil.ClientsetsInterface, itemResource *unstructured.Unstructured, resource, clusterNamespace string) error {
+	if resource == CephResourceCephClusters {
+		jsonPatchData, _ := json.Marshal(clusterResourcePatchFinalizer)
+		err := clientsets.PatchResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, itemResource.GetName(), types.MergePatchType, jsonPatchData)
+		if err != nil {
+			return err
+		}
+	}
+
+	jsonPatchData, _ := json.Marshal(defaultResourceRemoveFinalizers)
+	err := clientsets.PatchResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, itemResource.GetName(), types.MergePatchType, jsonPatchData)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func ensureClusterIsEmpty(ctx context.Context, k8sClientSet kubernetes.Interface, clusterNamespace string) {
+	logging.Info("waiting to clean up resources")
+	for {
+		pods, err := k8sClientSet.CoreV1().Pods(clusterNamespace).List(ctx, v1.ListOptions{LabelSelector: "rook_cluster=" + clusterNamespace})
```

Comment:
How about sleeping 5s between each check for pods that are still alive? Then we won't have such a tight loop to keep checking.